File: LeetCode/Problems/Python/852. Peak Index in a Mountain Array.py
#Binary search
#Time: O(logn)
#Space: O(1)
class Solution:
    def peakIndexInMountainArray(self, A: List[int]) -> int:
        if not A:
            return -1
        
        l = 0
        r = len(A) -1
        #we want to find the peek
        while l < r:
            mid = l + (r-l)//2
            #if mid is larger than it's right, mid can possibly be the peek
            #so we want to include it
            if A[mid] > A[mid+1]:
                r = mid
            #if mid is smaller, we are safe to say that mid is not the peek
            #we can take a step
            #when l == r, l is the peek
            else:
                l = mid+1
        return l
            

File: LeetCode/Problems/C++/852. Peak Index in a Mountain Array.cpp
//binary search
class Solution {
public:
    int peakIndexInMountainArray(vector<int>& A) {
        if(!A.size()) return -1;
        int N = A.size();
        int l = 0, r = N-1;
        
        while(l<r){
            int mid = l + (r-l)/2;
            if(A[mid]<A[mid+1]){
                l = mid+1;
            }
            else{
                r = mid;
            }
        }
        //cout << l << r;
        return r;
    }
};

//linear search
lass Solution {
public:
    int peakIndexInMountainArray(vector<int>& A) {
        if(!A.size()) return -1;
        
        for(int i=0;i<A.size();i++){
            if(A[i]>A[i+1]){
                return i;
            }
        }
        return 0;
    }
};